org-code-javabuilder/lib/src/main/java/org/code/javabuilder/JavabuilderTestExecutionListener.java

```diff
@@ -93,66 +90,64 @@ public void executionFinished(
-   * format of the error message is: "[error message] ([java class name]:[line number])"
+   * error message from the assertion will be included with key assertionError.
    *
-   * <p>If the test threw another exception, then format of the error message is: "[exception name]
-   * thrown at [java class name]:[line number]"
+   * <p>If the test threw a JavabuilderRuntimeError, the error details will include custom exception
+   * details generated by JavabuilderRuntimeError.
    *
-   * <p>For example:
-   *
-   * <p>assertion failed (MyTest:1)
-   *
-   * <p>FileNotFoundException thrown at MyTest:10
+   * <p>If the test threw any other exception, the error details will include the exception name and
+   * stack trace, with keys exceptionName and stackTrace.
    *
    * @param throwable Throwable thrown by the test
    * @param className name of the test class
    * @return the error message
    */
-  private String getErrorMessage(Throwable throwable, String className) {
-    final String errorLine = this.errorLine(throwable.getStackTrace(), className);
-    String errorMessage = null;
+  private HashMap<String, String> getErrorMessageDetails(Throwable throwable, String className) {
+    HashMap<String, String> errorDetails = new HashMap<>();
+    this.setErrorLineDetails(throwable.getStackTrace(), className, errorDetails);
 
-    // If there was an assertion failure, print the failure message. If a
-    // JavabuilderRuntimeException
-    // was thrown, print the fallback message if it has one. If it does not or if another exception
-    // was thrown, print the exception name.
+    // If there was an assertion failure, send on the message.
+    // If a JavabuilderRuntimeException was thrown, send the exception details in a hashmap.
+    // If another exception was thrown, send the exception name.
     if (throwable instanceof AssertionError) {
-      errorMessage = String.format("\t%s (%s)\n", throwable.getMessage(), errorLine);
+      errorDetails.put(ClientMessageDetailKeys.ASSERTION_ERROR, throwable.getMessage());
     } else if (throwable instanceof JavabuilderRuntimeException) {
       JavabuilderRuntimeException exception = (JavabuilderRuntimeException) throwable;
-      String fallbackMessage = exception.getFallbackMessage();
-      if (fallbackMessage != null) {
-        errorMessage = String.format("\t%s", fallbackMessage);
-      }
-    }
-    if (errorMessage == null) {
-      String additionalMessage = this.getAdditionalErrorMessage(throwable);
-      errorMessage =
-          String.format("\t%s thrown at %s\n", this.getThrowableDisplayName(throwable), errorLine);
-      if (additionalMessage != null) {
-        errorMessage = String.format("%s \t%s\n", errorMessage, additionalMessage);
-      }
+      errorDetails.putAll(exception.getExceptionDetails());
+      errorDetails.put(ClientMessageDetailKeys.TYPE, exception.getMessage());
```

Comment:
For my understanding, how is this different from what would be included in the `CAUSE_MESSAGE` key in the throwable util helper?

Reply:
The 'message' is the enum key we assign to each `JavabuilderRuntimeException`. When we send the exception as a `JavabuilderThrowableMessage` the key gets sent separately from the details, so we need to add it in here so we can get it on the Java Lab side.

Reply:
for example, the key is something like `FILE_NOT_FOUND`

Reply:
mmm not sure I'm following, I think I was asking a different question? I guess I'm confused because we already call `getMessage()` when you call `exception.getExceptionDetails()` and put it in the details hash under the `CAUSE_MESSAGE` key:

https://github.com/code-dot-org/javabuilder/blob/c05d467c9d4a8274383e5def7c6819881d3e8cf3/org-code-javabuilder/protocol/src/main/java/org/code/protocol/JavabuilderThrowableMessageUtils.java#L19-L54

Maybe the difference is that you're calling `getMessage()` on the `JavabuilderRuntimeException` itself, rather than its cause? Our exception hierarchy makes my brain hurt, so still trying to wrap (no pun intended) my head around it, so I'm probably just misunderstanding something...

Reply:
Yep the difference is that one is on the cause, the other on the exception itself. The cause can either be any throwable or null, depending on if [this constructor](https://github.com/code-dot-org/javabuilder/blob/c05d467c9d4a8274383e5def7c6819881d3e8cf3/org-code-javabuilder/protocol/src/main/java/org/code/protocol/JavabuilderRuntimeException.java#L16) is used or not (if it's not, the cause will be null). The message for the `JavabuilderRuntimeException` will always a short constant value. For example, if it's a `NeighborhoodRuntimeException` it will always be one of [these](https://github.com/code-dot-org/javabuilder/blob/c05d467c9d4a8274383e5def7c6819881d3e8cf3/org-code-javabuilder/neighborhood/src/main/java/org/code/neighborhood/ExceptionKeys.java#L3).

Reply:
Ahh got it got it, I was looking for an override for getMessage or something in JavabuilderRuntimeException, guess that's just default behavior for RuntimeException. Thanks for explaining!

![image](https://user-images.githubusercontent.com/25372625/154122721-1efc2e20-1fb3-4498-8a98-6c49a7b6684e.png)